### ISSUE TITLE
Load Balancer - Gzip compression

### ISSUE THREAD

**Issue by jloisel** [rancher/rancher#2840]
Our backend has gzip compression enabled. We use the Load balancer to have SSL termination. The load balancer servers the content as decompressed. It should server the content gzip compressed.

We are using Rancher v0.47.0. Same with v0.46.0. Tried with Load-balancer rancher/agent-instance:v0.5.0 and rancher/agent-instance:v0.6.0.

**Comment by ibuildthecloud** [comment]
@jloisel It's not obvious to me why this wouldn't work.  Are you absolutely sure your client you are testing with is sending the `Accept-Encoding: gzip, deflate` header?

**Comment by jloisel** [comment]
Hi Darren, yes the client is an AngularJS UI which sends `Accept-Encoding: gzip, deflate` in every request. We tested the same client in the following conditions:
- configured to hit backend on plain HTTP: receives gzipped content,
- configured to hit load-balancer with SSL certificate: receives decompressed content.

The load-balancer is configured to serve the content on port 443 with the configured SSL certificate (provided by a valid certification authority). Our backend is a Spring boot web application with embedded Tomcat 8.0.28, we enabled gzip compression with the `server.compression.enabled=true` property in application config.

**Comment by jloisel** [comment]
Gzip compression works fine in haproxy when enabling it by manually editing the haproxy.cfg within the container. I added the following lines in the default section:
```
 compression algo gzip
 compression type application/json
```

These settings are lost when restarting the container, as expected.

**Comment by alena1108** [comment]
@jloisel support for custom parameters in haproxy "defaults" section will get enabled in the next release; here is the PR:

https://github.com/rancher/cattle/pull/1081

**Comment by jloisel** [comment]
Awesome! :+1:

**Comment by vincent99** [comment]
This does not make any sense... The default behavior of haproxy is to pass on the `Accept-Encoding` header and if the backend server compresses it the compressed response is sent back to the client.  I just tested this with the current release and it works with both regular and SSL listeners.

By setting `compression algo gzip` you are forcing haproxy to compress and rewrite the response.  It is apparently not compressed coming out of your backend @jloisel, or it would be working already.  So there has to be something like Spring disables it based on one of the headers that vary when going through haproxy.

It is generally much more desirable to have your backend do the compression because it can be written out compressed in the first place instead of read and written again.  Also you can only have many backend containers but onlyone balancer per public host IP/port.

**Comment by jloisel** [comment]
@vincent99 You are right. We were comparing the behavior between staging and production environment which had different compression settings. Compression wasn't enabled on application/json in production.

There is a reason why we want to enable Gzip within HAProxy: the response time is much higher (130ms average) when enabling gzip on our backend than with in HAProxy (65ms). Gzip compression is much more efficient on LoadBalancer level. It's simply because HAProxy is a compiled C program while our backend is Java.

Gzip disabled on both Tomcat and HAProxy:
![gzip-disabled](https://cloud.githubusercontent.com/assets/3208716/11496327/46c95e88-9811-11e5-91f3-2712c997039d.png)

Gzip enabled on Tomcat backend only:
![gzip-tomcat](https://cloud.githubusercontent.com/assets/3208716/11496343/73729670-9811-11e5-9325-8c03d3a03d8b.png)

Finally, Gzip enabled in HAProxy only:
![gzip-haproxy](https://cloud.githubusercontent.com/assets/3208716/11496439/1aa8090c-9812-11e5-9033-24c852c00aa9.png)

Gzip compression with HAProxy is as fast as serving uncompressed content. On the other side, gzip compression with Tomcat is almost 25% slower (response time > 1sec vs 850ms). Of course, i've run each at least 10 times and took only the fastest one.

**Comment by y8** [comment]
With rancher-server v0.50 you can now just add

````
compression algo gzip
compression type text/html "text/html; charset=utf-8" text/html;charset=utf-8 text/plain text/css text/javascript application/x-javascript application/javascript application/ecmascript application/rss+xml application/atomsvc+xml application/atom+xml application/atom+xml;type=entry application/atom+xml;type=feed application/cmisquery+xml application/cmisallowableactions+xml application/cmisatom+xml application/cmistree+xml application/cmisacl+xml application/msword application/vnd.ms-excel application/vnd.ms-powerpoint image/svg+xml
````

To `global` section in LoadBalancer. And then compression will be enabled on all hosts by default. Check the mime types, so everything that is worth compression is compressed.

**Comment by jloisel** [comment]
Thanks, we'll do it when we upgrade to v0.50.0.

**Comment by deniseschannon** [comment]
Going to close as this is now available as of v0.50.0. @jloisel I'd recommend using the latest (v0.51.0) whenever you decide to upgrade. Or if you want to only upgrade to the v0.50.x series, please use v0.50.2 as it has some hot fixes.
